OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+                           const VkCommandBuffer* command_buffers, uint32_t count) {
+    absl::WriterMutexLock lock(&mutex_);
+    if (!pool_to_command_buffers_.contains(pool)) {
+      pool_to_command_buffers_[pool] = {};
+    }
+    absl::flat_hash_set<VkCommandBuffer>& associated_cbs = pool_to_command_buffers_.at(pool);
+    for (uint32_t i = 0; i < count; ++i) {
+      VkCommandBuffer cb = command_buffers[i];
+      associated_cbs.insert(cb);
+      command_buffer_to_device_[cb] = device;
+    }
+  }
+
+  void UntrackCommandBuffers(VkDevice device, VkCommandPool pool,
+                             const VkCommandBuffer* command_buffers, uint32_t count) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(pool_to_command_buffers_.contains(pool));
+    absl::flat_hash_set<VkCommandBuffer>& associated_command_buffers =
+        pool_to_command_buffers_.at(pool);
+    for (uint32_t i = 0; i < count; ++i) {
+      VkCommandBuffer command_buffer = command_buffers[i];
+      associated_command_buffers.erase(command_buffer);
+      CHECK(command_buffer_to_device_.contains(command_buffer));
+      CHECK(command_buffer_to_device_.at(command_buffer) == device);
+      command_buffer_to_device_.erase(command_buffer);
+    }
+    if (associated_command_buffers.empty()) {
+      pool_to_command_buffers_.erase(pool);
+    }
+  }
+
+  void MarkCommandBufferBegin(VkCommandBuffer command_buffer) {
+    // Even when we are not capturing we create state for this command buffer to allow the
+    // debug marker tracking.
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(!command_buffer_to_state_.contains(command_buffer));
+      command_buffer_to_state_[command_buffer] = {};
+    }
+    if (!IsCapturing()) {
+      return;
+    }
+
+    uint32_t slot_index = RecordTimestamp(command_buffer, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_state_.contains(command_buffer));
+      command_buffer_to_state_.at(command_buffer).command_buffer_begin_slot_index =
+          std::make_optional(slot_index);
+    }
+  }
+
+  void MarkCommandBufferEnd(VkCommandBuffer command_buffer) {
+    if (!IsCapturing()) {
+      return;
+    }
+
+    uint32_t slot_index = RecordTimestamp(command_buffer, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT);
+
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_state_.contains(command_buffer));
+      CommandBufferState& command_buffer_state = command_buffer_to_state_.at(command_buffer);
+      // Writing to this field is safe, as there can't be any operation on this command buffer
+      // in parallel.
+      command_buffer_state.command_buffer_end_slot_index = std::make_optional(slot_index);
+    }
+  }
+
+  void MarkDebugMarkerBegin(VkCommandBuffer command_buffer, const char* text,
+                            internal::Color color) {
+    CHECK(text != nullptr);
+    bool too_many_markers;
```

Comment:
The naming is not clear to me. This is about depth, not count, right? "marker_depth_exceeds_maximum"?